[AIRFLOW-2744] Allow RBAC to accept plugins for views and links.

oliviersm199 opened the pull request:
### Description

Airflow Users that wish to create plugins for the new www_rbac UI can not add plugin views or links. This PR fixes that by letting a user specify their plugins for www_rbac and maintains backwards compatibility with the existing plugins system.

Adds two additional members of an AirflowPlugin class to cover role based views and role based menu items. Specifies that the user can create a dictionary which has the metadata necessary to create a FlaskAppBuilder BaseView or MenuItem. This lets the plugins manager extend to the new www_rbac GUI.

### Tests

Adds unit tests under the plugins manager to test that the www_rbac views and menu links get added successfully.

### Documentation

Demonstrates in the plugins page how a user would create plugins for the new UI with an example.

Commit:
[AIRFLOW-2744] Allow RBAC to accept plugins for views and links.

Airflow Users that wish to create plugins for the new www_rbac UI
can not add plugin views or links. This PR fixes that by letting
a user specify their plugins for www_rbac and maintains backwards
compatibility with the existing plugins system.

ashb commented on airflow/plugins_manager.py:
```diff
@@ -49,6 +49,8 @@ class AirflowPlugin(object):
     flask_blueprints = []
     menu_links = []
+    role_based_views = []
+    role_based_menu_items = []
```
Lets call these `appbuilder_views` etc. The "role" is just the reason we switched to FAB.

oliviersm199 replied:
Changed it to use appbuilder_views etc.

ashb reviewed (commented):
Nice!

Commit:
Merge branch 'master' into enable-plugins-www-rbac

oliviersm199 committed:
Merge branch 'master' into enable-plugins-www-rbac

ashb commented on tests/plugins/test_plugin.py:
```diff
@@ -67,17 +68,39 @@ def test(self):
-    static_url_path='/static/test_plugin')
+    template_folder="templates",
+    static_folder="static",
+    static_url_path="/static/test_plugin")
```
Style nit: please remove un-related changes like this from the PR.

ashb commented on tests/plugins/test_plugin.py:
```diff
@@ -58,7 +59,7 @@ def plugin_macro():
 # Creating a flask admin BaseView
 class TestView(BaseView):
-    @expose('/')
+    @expose("/")
```
Style nit: please remove un-related changes like this from the PR.

ashb commented on docs/plugins.rst:
```diff
@@ -170,3 +191,16 @@ definitions in Airflow.
+Airflow 1.10 introduced role based views using FlaskAppBuilder. You can configure which UI is used by setting rbac = True.
+To support plugin blueprints for both versions while maintaining backwards compatibility, additional fields in the
+AirflowTestPlugin class were added for specifically adding views and links to FlaskAppBuilder.
+
```
Remove empty lines.

ashb commented on docs/plugins.rst:
```diff
@@ -170,3 +191,16 @@ definitions in Airflow.
+
+Airflow 1.10 introduced role based views using FlaskAppBuilder. You can configure which UI is used by setting rbac = True.
+To support plugin blueprints for both versions while maintaining backwards compatibility, additional fields in the
+AirflowTestPlugin class were added for specifically adding views and links to FlaskAppBuilder.
```
Please say here what the fields are.

ashb commented on docs/plugins.rst:
```diff
@@ -170,3 +191,16 @@ definitions in Airflow.
+
+
+Note on role based views
+-------
```
The `-----` should match the length of the line above otherwise RST will complain at us.

ashb replied:
This one still needs doing I think?

oliviersm199 replied:
Hey I thought I changed it but must have made a mistake with git. Please check again :)

ashb reviewed (changes requested):
Few minor nits then this looks good to merge.

Commit:
[AIRFLOW-2744] Allow RBAC to accept plugins for views and links.

Stylistic fixes from code review. Updated the docs to specify
exactly what new fields were added to plugins. Removed unrelated
changes.

Commit:
[AIRFLOW-2744] Allow RBAC to accept plugins for views and links.

Stylistic fixes from code review. Updated the docs to specify
exactly what new fields were added to plugins. Removed unrelated
changes.

ashb commented on tests/plugins/test_plugin.py:
```diff
@@ -67,17 +68,40 @@ def test(self):
 bp = Blueprint(
     "test_plugin", __name__,
-    template_folder='templates',  # registers airflow/plugins/templates as a Jinja template folder
+    template_folder='templates', # registers airflow/plugins/templates as a Jinja template folder
```
This double space is required by flake8.

Commit:
Merge branch 'enable-plugins-www-rbac' of https://github.com/1010data/incubator-airflow into enable-plugins-www-rbac
